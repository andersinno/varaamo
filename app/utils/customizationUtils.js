import constants from '../constants/AppConstants';

function getDefaultMunicipality() {
  const host = window.location.host;

  switch (host) {
    case 'hel':
      return 'helsinki';
    case 'espoo':
      return 'espoo';
    case 'vantaa':
      return 'vantaa';
    default:
      return undefined;
  }
}

function getCurrentCustomization() {
  const host = window.location.host;
  return constants.CUSTOMIZATIONS[host] || null;
}

function getCustomizationClassName() {
  switch (getCurrentCustomization()) {
    case 'ESPOO': {
      return 'espoo-customizations';
    }

    case 'VANTAA': {
      return 'vantaa-customizations';
    }

    case 'TAMPERE': {
      return 'tampere-customizations';
    }

    default: {
      return '';
    }
  }
}

export {
  getCurrentCustomization,
  getCustomizationClassName,
  getDefaultMunicipality,
};
