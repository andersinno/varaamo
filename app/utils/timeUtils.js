import constants from 'constants/AppConstants';
import { DEFAULT_SLOT_SIZE } from 'constants/SlotConstants';

import forEach from 'lodash/forEach';
import map from 'lodash/map';
import Moment from 'moment';
import { extendMoment } from 'moment-range';


const moment = extendMoment(Moment);

function addToDate(date, daysToIncrement) {
  const newDate = moment(date).add(daysToIncrement, 'days');

  return newDate.format(constants.DATE_FORMAT);
}

function getDateStartAndEndTimes(date, useTimeRange, startTime, endTime, duration) {
  if (!date) {
    return {};
  }
  const start = `${date}T00:00:00Z`;
  const end = `${date}T23:59:59Z`;
  if (useTimeRange && endTime && startTime) {
    const timeZone = moment().format('Z');
    const availableStart = `${date}T${startTime}:00${timeZone}`;
    const availableEnd = `${date}T${endTime}:00${timeZone}`;
    const availableBetween = `${availableStart},${availableEnd},${duration}`;
    return { availableBetween, end, start };
  }
  return { end, start };
}

function getDateString(date) {
  if (!date) {
    return moment().format(constants.DATE_FORMAT);
  }

  return date;
}

function getDuration(duration) {
  if (!duration) {
    return moment(constants.FILTER.timePeriod, constants.FILTER.timePeriodType).minutes();
  }
  return duration;
}

function getDurationHours(duration) {
  const value = duration || constants.FILTER.timePeriod;
  return value / 60;
}

function calculateDuration(duration, start, end) {
  const { timeFormat, timePeriodType } = constants.FILTER;
  const startTime = moment(start, timeFormat);
  const endTime = moment(end, timeFormat);
  const diffMinutes = endTime.diff(startTime, timePeriodType);
  return Math.min(duration, diffMinutes);
}

function getEndTimeString(endTime) {
  if (!endTime) {
    return '23:30';
  }
  return endTime;
}

function calculateEndTime(end, start) {
  const { timeFormat, timePeriod, timePeriodType } = constants.FILTER;
  const startTime = moment(start, timeFormat);
  const endTime = moment(end, timeFormat);
  if (startTime.isSameOrAfter(endTime)) {
    return startTime.add(timePeriod, timePeriodType).format(timeFormat);
  }
  return end;
}

function getStartTimeString(startTime) {
  if (!startTime) {
    const now = moment();
    const nextPeriod = moment()
      .startOf('hour')
      .add(constants.FILTER.timePeriod, constants.FILTER.timePeriodType);
    while (nextPeriod.isBefore(now)) {
      nextPeriod.add(constants.FILTER.timePeriod, constants.FILTER.timePeriodType);
    }
    return nextPeriod.format(constants.FILTER.timeFormat);
  }
  return startTime;
}

function getTimeSlots(
  start, end,
  period = DEFAULT_SLOT_SIZE,
  reservations = [],
  reservationsToEdit = []
) {
  if (!start || !end) {
    return [];
  }

  const range = moment.range(moment(start), moment(end));
  const duration = moment.duration(period);

  const reservationRanges = map(
    reservations, reservation => moment.range(
      moment(reservation.begin), moment(reservation.end)
    )
  );

  const editRanges = map(
    reservationsToEdit, reservation => moment.range(
      moment(reservation.begin), moment(reservation.end)
    )
  );

  const slots = map(
    Array.from(
      range.by(constants.FILTER.timePeriodType, {
        excludeEnd: true,
        step: duration.as(constants.FILTER.timePeriodType),
      })
    ),
    (startMoment) => {
      const endMoment = moment(startMoment).add(duration);
      const asISOString = `${startMoment.toISOString()}/${endMoment.toISOString()}`;
      const asString = `${startMoment.format(constants.TIME_FORMAT)}\u2013${endMoment.format(
        constants.TIME_FORMAT
      )}`;

      const slotRange = moment.range(startMoment, endMoment);
      const editing = editRanges.some(editRange => editRange.overlaps(slotRange));

      let reserved = false;
      let reservation = null;
      let reservationStarting = false;
      let reservationEnding = false;
      forEach(reservationRanges, (reservationRange, index) => {
        if (reservationRange.overlaps(slotRange)) {
          reserved = true;
          reservation = reservations[index];
          const [reservationStart, reservationEnd] = reservationRange.toDate();
          const [slotStart, slotEnd] = slotRange.toDate();
          reservationStarting = reservationStart.getTime() === slotStart.getTime();
          reservationEnding = reservationEnd.getTime() === slotEnd.getTime();
        }
      });

      return {
        asISOString,
        asString,
        editing,
        reservation,
        reservationStarting,
        reservationEnding,
        reserved,
        start: startMoment.toISOString(),
        end: endMoment.toISOString(),
      };
    }
  );

  return slots;
}

function isPastDate(date) {
  const now = moment();
  return moment(date).isBefore(now, 'day');
}

function prettifyHours(hours, showMinutes = false) {
  if (showMinutes && hours < 0.5) {
    const minutes = moment.duration(hours, 'hours').minutes();
    return `${minutes} min`;
  }

  const rounded = Math.ceil(hours * 2) / 2;
  return `${rounded} h`;
}

function padLeft(number) {
  return number < 10 ? `0${number}` : String(number);
}
/**
 * Convert time period to minutes;
 *
 * @param {string} period Time string, usually HH:MM:SS
 * @returns {Int} Period in minutes
 */
function periodToMinute(period) {
  return moment.duration(period).asMinutes();
}

/**
 * Get end time slot with minPeriod time range.
 * For example: start slot at 2AM, minPeriod = 1h, expected result 3AM
 *
 * @param {object} startSlot
 * @param {string} slotSize
 * @param {string} minPeriod
 * @return {object} endSlot
 */
function getEndTimeSlotWithMinPeriod(startSlot, minPeriod, slotSize) {
  const minPeriodInMinutes = periodToMinute(minPeriod) - periodToMinute(slotSize);
  // minPeriod always >= slotSize
  // minus 1 timeSlot here so the timediff between start slot and end slot is equal with minPeriod.

  return {
    resource: startSlot.resource,
    begin: moment(startSlot.begin).add(minPeriodInMinutes, 'minutes').toISOString(),
    end: moment(startSlot.end).add(minPeriodInMinutes, 'minutes').toISOString()
  };
}

/**
 * Get time different
 * This function can be use to compare time
 * @param {string} startTime ISO Time String
 * @param {string} endTime ISO Time String
 * @returns {int} timediff
 */
function getTimeDiff(startTime, endTime, unit) {
  return moment(startTime).diff(moment(endTime), unit);
}

export {
  addToDate,
  calculateDuration,
  calculateEndTime,
  getDateStartAndEndTimes,
  getDateString,
  getDuration,
  getDurationHours,
  getEndTimeString,
  getStartTimeString,
  getTimeSlots,
  isPastDate,
  prettifyHours,
  padLeft,
  periodToMinute,
  getEndTimeSlotWithMinPeriod,
  getTimeDiff
};
