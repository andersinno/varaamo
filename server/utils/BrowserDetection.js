import React from 'react';

/**
 * @description Check if user is using IE browser or old versions of Edge and if that's the case show browser unsupported messages
 *
 * @returns {JSX.Element}
 */
export default function BrowserDetection() {
  const scriptString = `

    var ua = window.navigator.userAgent;
    var msie = ua.indexOf('MSIE ');
    var trident = ua.indexOf('Trident/');
    // 'Edge' is the user agent for old version of Edge
    // In new versions of Edge user agent is 'Edg'
    var edge = ua.indexOf('Edge');

    if (msie > 0 || trident > 0 || edge > 0) {

      var style = 'background-color: #ffec77; padding: .75rem 1.25rem; margin-bottom: 1rem;' +
      'border: 1px solid transparent; font-size: 16px; font-family: "Times"';

      var eng = 'Varaamo doesn’t support Internet Explorer any more. Please use a ' +
      'newer browser such as <a href="https://www.google.com/chrome/"> Chrome</a>, ' +
      '<a href="https://www.mozilla.org/en-US/firefox/new/"> Firefox </a> or ' +
      '<a href="https://www.microsoft.com/en-us/windows/microsoft-edge"> Edge</a>.';

      var fin = 'Varaamo ei tue enää Internet Explorer -selainta. Käytä jotain ' +
      'uudempaa selainta, kuten <a href="https://www.google.com/chrome/"> Chrome</a>, ' +
      '<a href="https://www.mozilla.org/en-US/firefox/new/"> Firefox </a> tai ' +
      '<a href="https://www.microsoft.com/en-us/windows/microsoft-edge"> Edge</a>.';

      var swe = 'Varaamo fungerar inte längre med Internet Explorer. Så var vänlig' +
      'och använd en nyare  browser som <a href="https://www.google.com/chrome/"> Chrome</a>,' +
      '<a href="https://www.mozilla.org/en-US/firefox/new/"> Firefox </a> eller' +
      '<a href="https://www.microsoft.com/en-us/windows/microsoft-edge"> Edge</a>.';

      var array = [fin, eng, swe];
      array.forEach(function (text, index) {
          var tag = document.createElement("p");
          tag.innerHTML = text;
          tag.setAttribute("style", style);
          document.getElementById("root").appendChild(tag);
      });
    }
  `;

  // eslint-disable-next-line react/react-in-jsx-scope
  return <script dangerouslySetInnerHTML={{ __html: scriptString }} />;
}
