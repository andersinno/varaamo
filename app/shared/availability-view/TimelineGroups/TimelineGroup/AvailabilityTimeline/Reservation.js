import classNames from 'classnames';
import moment from 'moment';
import PropTypes from 'prop-types';
import React from 'react';
import Glyphicon from 'react-bootstrap/lib/Glyphicon';
import OverlayTrigger from 'react-bootstrap/lib/OverlayTrigger';
import Popover from 'react-bootstrap/lib/Popover';

import ReservationAccessCode from 'shared/reservation-access-code';
import utils from '../utils';

function getReserverName(reserverName, user) {
  return reserverName || (user && (user.displayName || user.email));
}

Reservation.propTypes = {
  accessCode: PropTypes.string,
  begin: PropTypes.string.isRequired,
  comments: PropTypes.string,
  end: PropTypes.string.isRequired,
  eventSubject: PropTypes.string,
  id: PropTypes.number.isRequired,
  numberOfParticipants: PropTypes.number,
  onClick: PropTypes.func,
  reserverName: PropTypes.string,
  state: PropTypes.string,
  user: PropTypes.shape({
    displayName: PropTypes.string,
    email: PropTypes.string,
  }),
};

function Reservation({ onClick, ...reservation }) {
  const startTime = moment(reservation.begin);
  const endTime = moment(reservation.end);
  const width = utils.getTimeSlotWidth({ startTime, endTime });
  const reserverName = getReserverName(reservation.reserverName, reservation.user);
  const popover = (
    <Popover className="reservation-info-popover" id={`popover-${reservation.id}`} title={reservation.eventSubject}>
      <div>
        <Glyphicon glyph="time" />
        {`${startTime.format('HH:mm')} - ${endTime.format('HH:mm')}`}
      </div>
      {reserverName && <div>{reserverName}</div>}
      {reservation.numberOfParticipants && (
      <div>
        <Glyphicon glyph="user" />
        {' '}
        {reservation.numberOfParticipants}
      </div>
      )}
      {reservation.accessCode && <div><ReservationAccessCode reservation={reservation} /></div>}
      {reservation.comments && <hr />}
      {reservation.comments && (
      <div>
        <Glyphicon glyph="comment" />
        {' '}
        {reservation.comments}
      </div>
      )}
    </Popover>
  );
  return (
    <button
      className={classNames('reservation-link', { 'with-comments': reservation.comments })}
      onClick={() => onClick && reservation.userPermissions.canModify && onClick(reservation)}
      type="button"
    >
      <OverlayTrigger
        overlay={popover}
        placement="top"
        trigger={['hover', 'focus']}
      >
        <span
          className={classNames('reservation',
            { requested: reservation.state === 'requested' },
            { disabled: reservation.state === 'confirmed' && !reservation.isOwn && !reservation.userPermissions.canModify },
            { reserved: reservation.state === 'confirmed' && !reservation.isOwn && reservation.userPermissions.canModify })}
          style={{ width }}
        >
          <span className="names">
            <span className="event-subject">{reservation.eventSubject}</span>
            <span className="reserver-name">{reserverName}</span>
          </span>
        </span>
      </OverlayTrigger>
    </button>
  );
}

export default Reservation;
