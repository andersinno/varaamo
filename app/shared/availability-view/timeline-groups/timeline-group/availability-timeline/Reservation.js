import classNames from 'classnames';
import moment from 'moment';
import PropTypes from 'prop-types';
import React from 'react';
import Glyphicon from 'react-bootstrap/lib/Glyphicon';
import OverlayTrigger from 'react-bootstrap/lib/OverlayTrigger';
import Popover from 'react-bootstrap/lib/Popover';

import { getReservationPrice, getHasOnlinePaymentSupport } from '../../../../../../src/domain/resource/utils';
import ReservationAccessCode from '../../../../reservation-access-code/ReservationAccessCode';
import utils from '../utils';

function getReserverName(reserverName, user) {
  return reserverName || (user && (user.displayName || user.email));
}

Reservation.propTypes = {
  accessCode: PropTypes.string,
  begin: PropTypes.string.isRequired,
  comments: PropTypes.string,
  end: PropTypes.string.isRequired,
  eventSubject: PropTypes.string,
  id: PropTypes.number.isRequired,
  numberOfParticipants: PropTypes.number,
  onClick: PropTypes.func,
  reserverName: PropTypes.string,
  products: PropTypes.array,
  state: PropTypes.string,
  user: PropTypes.shape({
    displayName: PropTypes.string,
    email: PropTypes.string,
  }),
};

function Reservation({ onClick, ...reservation }) {
  const startTime = moment(reservation.begin);
  const endTime = moment(reservation.end);
  const width = utils.getTimeSlotWidth({ startTime, endTime });
  const reserverName = getReserverName(reservation.reserverName, reservation.user);
  // This value is zero when no price is found.
  const price = getReservationPrice(reservation.begin, reservation.end, { products: reservation.products });
  const showPrice = getHasOnlinePaymentSupport(reservation);

  const popover = (
    <Popover className="reservation-info-popover" id={`popover-${reservation.id}`} title={reservation.eventSubject}>
      <div>
        <Glyphicon glyph="time" />
        {`${startTime.format('HH:mm')} - ${endTime.format('HH:mm')}`}
      </div>
      {reserverName && <div>{reserverName}</div>}
      {reservation.numberOfParticipants && (
        <div>
          <Glyphicon glyph="user" />
          {' '}
          {reservation.numberOfParticipants}
        </div>
      )}
      {reservation.accessCode && <div><ReservationAccessCode reservation={reservation} /></div>}
      {reservation.comments && <hr />}
      {reservation.comments && (
        <div>
          <Glyphicon glyph="comment" />
          {' '}
          {reservation.comments}
        </div>
      )}
    </Popover>
  );

  return (
    <button
      className={classNames('reservation-link', { 'with-comments': reservation.comments })}
      onClick={() => onClick && reservation.userPermissions.canModify && onClick(reservation)}
      type="button"
    >
      <OverlayTrigger
        overlay={popover}
        placement="top"
        trigger={['hover', 'focus']}
      >
        <span
          className={classNames('reservation',
            {
              requested: reservation.state === 'requested',
            },
            {
              disabled: reservation.state === 'confirmed'
                && !reservation.isOwn
                && !reservation.userPermissions.canModify,
            },
            {
              reserved: reservation.state === 'confirmed'
                && !reservation.isOwn
                && reservation.userPermissions.canModify,
            },
            {
              waiting: reservation.state === 'waiting_for_payment',
            },
            {
              paid: reservation.state === 'confirmed'
                && !reservation.staffEvent
                && price,
            })}
          style={{ width }}
        >
          <span className="names">
            {reservation.eventSubject && <span className="event-subject">{reservation.eventSubject}</span>}
            <span className="reserver-name">{reserverName}</span>
            {showPrice && !reservation.staffEvent && (
              <span className="price-tag" data-testid="price">{`${price} â‚¬`}</span>
            )}
          </span>
        </span>
      </OverlayTrigger>
    </button>
  );
}

export default Reservation;
