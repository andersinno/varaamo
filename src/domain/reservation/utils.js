import moment from 'moment';
import find from 'lodash/find';
import sortBy from 'lodash/sortBy';
import clone from 'lodash/clone';
import tail from 'lodash/tail';
import last from 'lodash/last';
import merge from 'lodash/merge';
import get from 'lodash/get';

import client from '../../common/api/client';
import { RESERVATION_STATE } from '../../constants/ReservationState';
import { getReservationPrice } from '../resource/utils';

export const combineReservations = (reservations) => {
  if (!reservations || !reservations.length) {
    return [];
  }

  const sorted = sortBy(reservations, 'begin');
  const initialValue = [clone(sorted[0])];

  return tail(sorted).reduce((previous, current) => {
    if (current.begin === last(previous).end) {
      last(previous).end = current.end;
    } else {
      previous.push(clone(current));
    }
    return previous;
  }, initialValue);
};

export const getNextAvailableTime = (reservations, fromMoment = moment()) => {
  const combinedReservations = combineReservations(reservations);
  if (!combinedReservations.length || fromMoment < moment(combinedReservations[0].begin)) {
    return fromMoment;
  }

  const ongoingReservation = find(combinedReservations, reservation => (
    moment(reservation.begin) <= fromMoment && fromMoment < moment(reservation.end)
  ));

  return ongoingReservation ? moment(ongoingReservation.end) : fromMoment;
};

export const getCurrentReservation = (reservations) => {
  const now = moment();
  return find(reservations, (reservation) => {
    return moment(reservation.begin) < now && now < moment(reservation.end);
  });
};

/**
 * Edit existing reservation API helper
 * @param {Object} reservation Original reservation data
 * @param {Object} fields Edit fields object
 * @memberof ManageReservationsPage
 * @returns {Promise}
 */
export const putReservation = (reservation, fields) => {
  return client.put(`reservation/${reservation.id}`, merge(reservation, fields), { include: 'resource_detail' });
};
/**
 * Delete/Cancel reservation
 *
 * @param {Object} reservation
 * @returns {Promise}
 */
export const cancelReservation = (reservation) => {
  return client.delete(`reservation/${reservation.id}`);
};

/**
 * Generate reservation edit URL from reservation data.
 *
 * @param {*} reservation
 * @returns {string} Reservation URL
 */
export const getEditReservationUrl = (reservation) => {
  const {
    begin, end, id, resource,
  } = reservation;
  const date = moment(begin).format('YYYY-MM-DD');
  const beginStr = moment(begin).format('HH:mm');
  const endStr = moment(end).format('HH:mm');

  return `/reservation?begin=${beginStr}&date=${date}&end=${endStr}&id=${id || ''}&resource=${resource.id}`;
};

/**
 * Check if current user (logged in user) have
 * permission to modify selected reservation.
 *
 * Reservation which in canceled state can not be changed to something else.
 *
 * @param {Object} reservation
 * @returns {Boolean} False by default
 */
export const canUserModifyReservation = (reservation) => {
  if (get(reservation, 'user_permissions.can_modify', false)
      && reservation.state !== RESERVATION_STATE.CANCELLED) {
    return true;
  }

  return false;
};

/**
 * Check if current user (logged in user) have
 * permission to cancel (delete) selected reservation.
 *
 * Reservation which in canceled state can not be changed to something else.
 *
 * @param {Object} reservation
 * @returns {Boolean} False by default
 */
export const canUserCancelReservation = (reservation) => {
  if (get(reservation, 'user_permissions.can_delete', false)
      && reservation.state !== RESERVATION_STATE.CANCELLED) {
    return true;
  }

  return false;
};

/**
 * Check if refund policy should be shown.
 *
 * @param {Boolean} isAdmin
 * @param {Object} reservation
 * @param {Object} resource
 * @returns {Boolean}
 */
export const getShowRefundPolicy = (isAdmin, reservation, resource) => {
  const isStaffEvent = get(reservation, 'staffEvent', false);
  const price = getReservationPrice(reservation.begin, reservation.end, resource);

  return isAdmin && !isStaffEvent && price > 0;
};
