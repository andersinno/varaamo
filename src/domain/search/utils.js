import get from 'lodash/get';
import range from 'lodash/range';
import capitalize from 'lodash/capitalize';
import sortBy from 'lodash/sortBy';
import snakeCase from 'lodash/snakeCase';
import forEach from 'lodash/forEach';
import moment from 'moment';

import { getDefaultMunicipality } from '../../../app/utils/customizationUtils';
import constants from '../../../app/constants/AppConstants';
import * as urlUtils from '../../common/url/utils';
import settings from '../../../config/settings';

export const getFiltersFromUrl = (location, supportedFilters = constants.SUPPORTED_SEARCH_FILTERS) => {
  const query = new URLSearchParams(location.search);
  const defaultDate = moment().format(constants.DATE_FORMAT);
  const defaultMunicipality = getDefaultMunicipality();

  const filters = {
    // Give default date to populate start/end time as default when fetching resources,
    // Otherwise, reservations property under resource data will be
    // empty.
    date: defaultDate,
  };

  if (defaultMunicipality) {
    // Determine current version of Varaamo (Helsinki, Espoo, Vantaa)
    // and filter results to target that municipality by default.
    filters.municipality = defaultMunicipality;
  }

  query.forEach((value, key) => {
    if (!supportedFilters || supportedFilters[key] !== undefined) {
      filters[key] = value;
    }
  });

  return filters;
};

export const getSearchFromFilters = (filters) => {
  return urlUtils.getSearch(filters);
};

export const getApiParamsFromFilters = (filters) => {
  const params = {};
  forEach(filters, (value, key) => {
    params[snakeCase(key)] = value;
  });

  return params;
};

export const getUnitOptions = (units, locale) => {
  const options = units.map((unit) => {
    const finnishName = get(unit, 'name.fi');

    return ({
      value: unit.id,
      // Use name in Finnish in case it doesn't exist for current locale
      label: get(unit, `name[${locale}]`, finnishName),
    });
  });

  return sortBy(options, 'label');
};

export const getPurposeOptions = (purposes, locale) => {
  const options = purposes
    .filter(purpose => purpose.parent === null)
    .map(unit => ({
      value: unit.id,
      label: get(unit, `name[${locale}]`, ''),
    }));

  return sortBy(options, 'label');
};

/**
 * Getter for people capacity options.
 * @returns {{label: number, value: number}[]}
 */
export const getPeopleCapacityOptions = () => {
  return [
    ...range(1, 10),
    ...range(10, 35, 5),
    ...range(40, 110, 10),
  ].map(number => ({ label: number, value: number }));
};

export const getClosestPeopleCapacityValue = (value) => {
  return getPeopleCapacityOptions()
    .map(option => option.value)
    .reduce((previous, current) => {
      if (Math.abs(current - value) < Math.abs(previous - value) && current - value < 1) {
        return current;
      }

      return previous;
    });
};

/**
 * Getter for municipality options.
 * @returns {{label: string, value: string}[]}
 */
export const getMunicipalityOptions = () => {
  let municipalities = constants.DEFAULT_MUNICIPALITY_OPTIONS;

  if (
    Array.isArray(settings.CUSTOM_MUNICIPALITY_OPTIONS)
    && settings.CUSTOM_MUNICIPALITY_OPTIONS.length
  ) {
    municipalities = settings.CUSTOM_MUNICIPALITY_OPTIONS;
  }

  return municipalities.map((municipality) => {
    const municipalityStr = typeof municipality === 'string' ? municipality : municipality.toString();

    return {
      value: municipalityStr.toLowerCase(),
      label: capitalize(municipalityStr),
    };
  });
};

/**
 * Getter for search page url.
 * Just give it the filters and it returns you the search page path.
 * @param filters {object}
 * @returns {string}
 */
export const getSearchPageLink = (filters) => {
  return urlUtils.getLinkString('/search', getSearchFromFilters(filters));
};
