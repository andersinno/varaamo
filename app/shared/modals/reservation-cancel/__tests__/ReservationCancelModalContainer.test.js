import React from 'react';
import Button from 'react-bootstrap/lib/Button';
import Modal from 'react-bootstrap/lib/Modal';
import simple from 'simple-mock';
import toJSON from 'enzyme-to-json';

import CompactReservationList from '../../../compact-reservation-list/CompactReservationList';
import Reservation from '../../../../utils/fixtures/Reservation';
import Resource from '../../../../utils/fixtures/Resource';
import { shallowWithIntl } from '../../../../utils/testUtils';
import {
  UnconnectedReservationCancelModalContainer as ReservationCancelModalContainer,
} from '../ReservationCancelModalContainer';

describe('shared/modals/reservation-cancel/ReservationCancelModalContainer', () => {
  const resource = Resource.build();
  const reservation = Reservation.build({ resource: resource.id });
  const defaultProps = {
    actions: {
      closeReservationCancelModal: () => null,
      deleteReservation: () => null,
    },
    cancelAllowed: false,
    isCancellingReservations: false,
    reservation,
    resource,
    show: true,
  };

  function getWrapper(extraProps = {}) {
    return shallowWithIntl(<ReservationCancelModalContainer {...defaultProps} {...extraProps} />);
  }

  describe('render', () => {
    test('renders a Modal component', () => {
      const modalComponent = getWrapper().find(Modal);
      expect(modalComponent.length).toBe(1);
    });

    describe('Modal header', () => {
      function getModalHeaderWrapper(props) {
        return getWrapper(props).find(Modal.Header);
      }

      test('is rendered', () => {
        expect(getModalHeaderWrapper()).toHaveLength(1);
      });

      test('contains a close button', () => {
        expect(getModalHeaderWrapper().props().closeButton).toBe(true);
      });

      describe('title', () => {
        test('is correct if cancel is allowed', () => {
          const modalTitle = getModalHeaderWrapper({ cancelAllowed: true }).find(Modal.Title);
          expect(modalTitle.length).toBe(1);
          expect(modalTitle.prop('children')).toBe('ReservationCancelModal.cancelAllowedTitle');
        });

        test('is correct if cancel is not allowed', () => {
          const modalTitle = getModalHeaderWrapper({ cancelAllowed: false }).find(Modal.Title);
          expect(modalTitle.length).toBe(1);
          expect(modalTitle.prop('children')).toBe('ReservationCancelModal.cancelNotAllowedTitle');
        });
      });
    });

    describe('Modal body', () => {
      function getModalBodyWrapper(props) {
        return getWrapper(props).find(Modal.Body);
      }

      test('is rendered', () => {
        expect(getModalBodyWrapper).toHaveLength(1);
      });

      describe('if cancel is allowed', () => {
        const cancelAllowed = true;

        test('renders CompactReservationList', () => {
          expect(
            getModalBodyWrapper({ cancelAllowed }).find(CompactReservationList),
          ).toHaveLength(1);
        });

        test('does not render responsibleContactInfo', () => {
          expect(
            getModalBodyWrapper({ cancelAllowed }).find('.responsible-contact-info'),
          ).toHaveLength(0);
        });
      });

      describe('if cancel is not allowed', () => {
        const cancelAllowed = false;

        test('renders responsibleContactInfo', () => {
          expect(
            getModalBodyWrapper({ cancelAllowed }).find('.responsible-contact-info'),
          ).toHaveLength(1);
        });

        test('does not render CompactReservationList', () => {
          expect(
            getModalBodyWrapper({ cancelAllowed }).find(CompactReservationList),
          ).toHaveLength(0);
        });
      });

      describe('if resource has payment support', () => {
        test('renders custom message', () => {
          const wrapper = getModalBodyWrapper({
            resource: {
              products: [
                {
                  type: 'rent',
                },
              ],
            },
          });

          expect(toJSON(wrapper)).toMatchSnapshot();
        });
      });
    });

    describe('Footer buttons', () => {
      function getFooterButtonsWrapper(props) {
        return getWrapper(props).find(Modal.Footer).find(Button);
      }

      describe('if cancel is allowed', () => {
        const cancelAllowed = true;
        const buttons = getFooterButtonsWrapper({ cancelAllowed });

        test('renders cancel button', () => {
          expect(buttons.at(0).props().children).toBe('ReservationCancelModal.cancelAllowedCancel');
        });

        test('renders confirm button', () => {
          expect(buttons.at(1).props().children).toBe('ReservationCancelModal.cancelAllowedConfirm');
        });
      });

      describe('if cancel is not allowed', () => {
        const cancelAllowed = false;
        const buttons = getFooterButtonsWrapper({ cancelAllowed });

        test('renders back button', () => {
          expect(buttons.at(0).props().children).toBe('common.back');
        });
      });
    });
  });

  describe('handleCancel', () => {
    const closeReservationCancelModal = simple.mock();
    const deleteReservation = simple.mock();
    const actions = { closeReservationCancelModal, deleteReservation };

    beforeAll(() => {
      const instance = getWrapper({ actions }).instance();
      instance.handleCancel();
    });

    test('calls deleteReservation with the reservation', () => {
      expect(deleteReservation.callCount).toBe(1);
      expect(deleteReservation.lastCall.arg).toEqual(defaultProps.reservation);
    });

    test('closes the ReservationCancelModal', () => {
      expect(closeReservationCancelModal.callCount).toBe(1);
    });
  });
});
