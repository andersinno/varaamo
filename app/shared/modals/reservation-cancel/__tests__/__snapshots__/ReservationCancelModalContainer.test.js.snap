// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shared/modals/reservation-cancel/ReservationCancelModalContainer render Modal body if resource has payment support renders custom message 1`] = `
<ModalBody
  bsClass="modal-body"
  componentClass="div"
>
  <div>
    <p>
      ReservationCancelModal.cancelNotAllowedInfoOnlinePayment.1
    </p>
    <p>
      ReservationCancelModal.cancelNotAllowedInfoOnlinePayment.2
    </p>
    <p
      className="responsible-contact-info"
    />
  </div>
</ModalBody>
`;
